c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -723,8 +732,9 @@ pub fn visit<'tcx>(
         equiv_constraints: Vec::new(),
     };
 
-    for (ptr, perms) in acx.string_literal_perms() {
+    for (ptr, perms, neg_perms) in acx.string_literal_perms() {
         tc.constraints.add_all_perms(ptr, perms);
+        tc.constraints.add_no_perms(ptr, neg_perms);
```

Comment:
There might be a subtle answer I'm missing: what is the difference between "remove permissions", "add no permissions", and "add forbidden permissions"?

Reply:
The various `add_foo` methods each add a `foo` constraint to the constraint set.  `add_no_perms(ptr, perms)` adds `Constraint::NoPerms(ptr, perms)`, which requires that `ptr` has none of the permissions listed in `perms`.

Maybe "emit" would be clearer than "add" for these methods?

Reply:
`emit` usually makes me think of code generation, so it's not much better.

I think the distinction between "remove permissions" and "add no permissions" is clearer, but is there one between "add no permissions" and "updates forbidden"?

Reply:
`Constraint::NoPerms` is generated when we encounter code that implies that a pointer must not have a certain permission.  For example, on `p = ptr::null()`, we emit (approximately) `Constraint::NoPerms(p, NON_NULL)` - `p` must not have the `NON_NULL` permission.  This is mainly relevant for permissions that are set by default on all pointers and are removed only when required by specific constraints (`NON_NULL` is one such permission).

"Updates forbidden" is not a constraint; instead, it causes the initial value of a permission to be preserved, ignoring any constraints that would require otherwise.  For example, if we have `Constraint::NoPerms(p, NON_NULL)`, but `updates_forbidden[p]` contains `NON_NULL`.  Normally, processing that constraint during `DataflowConstraints::propagate` would remove the `NON_NULL` permission from `p`.  But because `NON_NULL` is listed in `updates_forbidden[p]`, the change to `p`'s `NON_NULL` flag is suppressed.